aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp

```diff
@@ -1249,6 +1323,124 @@ magma_trans_t to_magma(TransposeType trans) {
   auto storage_##name = pin_memory<type>(size); \
   name = static_cast<type*>(storage_##name.data());
 
+namespace {
+
+template <typename scalar_t>
+void apply_ldl_factor(
+    const Tensor& A,
+    const Tensor& pivots,
+    const Tensor& info,
+    bool upper) {
+#if !AT_MAGMA_ENABLED()
+  TORCH_CHECK(
+      false,
+      "torch.linalg.ldl_factor: MAGMA library not found in "
+      "compilation. Please rebuild with MAGMA.");
+#else
+  auto batch_size = batchCount(A);
+  magma_int_t n = magma_int_cast(A.size(-2), "A.size(-2)");
+  magma_int_t leading_dim = magma_int_cast(A.stride(-1), "A.stride(-1)");
+  magma_uplo_t uplo = upper ? MagmaUpper : MagmaLower;
+
+  auto a_stride = A.dim() > 2 ? A.stride(-3) : 0;
+  auto pivots_stride = pivots.dim() > 1 ? pivots.stride(-2) : 0;
+
+  auto a_data = A.data_ptr<scalar_t>();
+  Tensor pivots_cpu =
+      at::empty_like(pivots, pivots.options().device(kCPU).pinned_memory(true));
+  auto pivots_data = pivots_cpu.data_ptr<magma_int_t>();
+  Tensor info_cpu =
+      at::empty_like(info, info.options().device(kCPU).pinned_memory(true));
+  auto info_data = info_cpu.data_ptr<magma_int_t>();
```

Comment:
We don't need to create all the tensors. We can simply reserve enough memory for the pivot of one matrix (`std::vector` should do for that) and then copy them to the right position in pivots and info in each iteration of the loop (flattening the few dimensions via a `view`, for example)

Reply:
I prefer using `at::Tensor` for reserving contiguous blocks of memory on the CPU.

Reply:
I prefer using `at::Tensor` for reserving contiguous blocks of memory on the CPU instead of `std::vector` and we need it to be pinned for performance.
I don't think it's necessarily faster to reserve memory for one matrix/pivot in a batch and copy it one-by-one back to GPU when the result is computed.

Reply:
This is what we're currently doing in `cusovler_gesvd`, so we should figure out what's the best way of doing this and do it consistently across the code base.
A priori, I'd say that if they are in pinned memory and all these things are async (the copies can be async, as long as you synchronise at the end) it should be better not to reserve the full batch of matrices but just the one matrix of the correct size but idk. It's not that important really.